runtime/compiler/optimizer/InterpreterEmulator.cpp

```diff
@@ -43,7 +251,15 @@ InterpreterEmulator::maintainStackForIf(TR_J9ByteCode bc)
    IconstOperand * first = pop()->asIconst();
    bool canBranch = true;
    bool canFallThru = true;
-   if (second && first)
+   // Comment out the branch folding as all the paths have to be interpreted in order
+   // to propagate object info in operand stack or local slots. Since branch folding
+   // currently only affects thunk archetypes, with similar branch folding in ilgen,
+   // calls in dead path won't be inlined, disabling the following code doesn't affect
+   // performance
+   // TODO: add code to record dead path and ignore it in object info propagation, enable
+   // the following code if branch folding is possible in LambdaForm methods
+   //
+   if (false && second && first)
```

Comment:
@vijaysun-omr Here's the first change to attain the performance. The change in this PR will propagate state in local slots and operand stack in bytecode iterator's order. A block will get local slots and operand stack states from its predecessors, but this requires all the predecessors to be visited before the successor, otherwise we set the local slots and operand stack state of the block being visited to unknown, this will cause the failure to find a call target for a MethodHandle call as the target searching requires known MethodHandle object as the receiver of the call.

The disabled code here does branch folding, so we may skip interpreting a block's bytecodes, introducing blocks that are never visited, stopping the propagation of object info to their successors.

Reply:
I see, thanks for the good explanation. That makes sense but not very obvious until you clarified :)